File: docs/validation/cpp/app/src/scenarios/s02_read_binary.cpp
// S-02: Read reference fixture with binary data (correction grids)
#include "s02_read_binary.hpp"
#include "machine_config/machine_config.hpp"

#include <algorithm>
#include <cmath>

namespace s02 {

using namespace machine_config;

scenarios::Result run(const std::filesystem::path& fixturesDir, const std::filesystem::path&) {
    auto path = fixturesDir / "reference_config.h5";
    try {
        MachineConfigReader reader{path};
        CorrectionData cd = reader.getCorrectionData(0);
        CorrectionData icd = reader.getInverseCorrectionData(0);

        if (cd.shape != std::array<std::size_t, 3>{257, 257, 2}) {
            return {false, "correction_data shape mismatch"};
        }
        if (icd.shape != std::array<std::size_t, 3>{257, 257, 2}) {
            return {false, "inverse_correction_data shape mismatch"};
        }
        if (!std::any_of(cd.data.begin(), cd.data.end(), [](double v) { return std::isfinite(v); })) {
            return {false, "correction_data: no finite values"};
        }
        if (!std::any_of(icd.data.begin(), icd.data.end(), [](double v) { return std::isfinite(v); })) {
            return {false, "inverse_correction_data: no finite values"};
        }
        if (scenarios::bitwiseEqual(cd.data, icd.data)) {
            return {false, "correction_data and inverse_correction_data are identical"};
        }

        std::string hash = scenarios::sha256Hex(cd.data);
        return {true, "shapes OK, finite values OK, forward!=inverse, correction_data SHA-256=" + hash};
    } catch (const std::exception& e) {
        return {false, std::string("exception: ") + e.what()};
    }
}

} // namespace s02

File: docs/validation/go/app/scenarios/s02_read_binary.go
package scenarios

// S-02: Read correction grids, shape and finite values
//
// ID:       S-02
// Action:   GetCorrectionData(0) on reference_config.h5.
// Expected: shape == [257,257,2]; grid contains NaNs (out-of-field cells).

import (
	"fmt"
	"math"
	"path/filepath"

	mc "machine-config-go"
)

func RunS02ReadBinary(fixturesDir, _ string) (bool, string) {
	path := filepath.Join(fixturesDir, "reference_config.h5")
	cd, err := mc.NewReader(path).GetCorrectionData(0)
	if err != nil {
		return false, fmt.Sprintf("GetCorrectionData failed: %v", err)
	}
	if cd.Shape != [3]int{257, 257, 2} {
		return false, fmt.Sprintf("shape = %v", cd.Shape)
	}
	nan := 0
	for _, v := range cd.Data {
		if math.IsNaN(v) {
			nan++
		}
	}
	if nan == 0 {
		return false, "expected NaNs in correction grid"
	}
	return true, fmt.Sprintf("shape=%v nan_count=%d total=%d", cd.Shape, nan, len(cd.Data))
}
